### ISSUE TITLE
Allow injecting code into the import routine

### ISSUE THREAD

**Issue by djerius** [preaction/Import-Base#21]
It'd be very helpful if one could specify code to be run as part of the `import()` process.

For example, as part of a bundle, I'd like to apply a role to the importing package.  It's doable by creating an import routine in the subclass of `Import::Base`, calling the parent import, then applying the role, but the caller level would then be messed up (and I'd have to figure out how to fix that).

As for the API, perhaps using a pseudomodule (call it `import_inject`):
```
 import_inject => sub { my ( $target ) = @_; Moo::Role->apply_roles_to_package( $target ) }
```
Thanks,

Diab

**Comment by djerius** [comment]
I've a kludgy, but trivial work around:  create a module which applies roles when imported, e.g.:
```
package ApplyRole;

use Moo::Role;

sub import {
    my  $class = shift;
    my $target = caller;

    Moo::Role->apply_role_to_package( $target, @_ );

}
1;
```
It'd be nice to avoid the extra package

**Comment by preaction** [comment]
Did you try adding a subref to `@IMPORT_MODULES` or `%IMPORT_BUNDLES`? I do something similar, but perhaps in a bit of bad way, here: https://github.com/preaction/Bot-Freyr/blob/master/lib/Bot/Freyr/Base.pm#L42

**Comment by djerius** [comment]
Ahh.  Thanks, I missed the fact that there's an `$args->{package}`.  That should do nicely.

Sorry for the noise.

Diab

**Comment by preaction** [comment]
I'll add an example to improve the docs there. I remember deliberately minimizing the documentation for that because I wasn't sure if I liked the idea, but it's useful (as you've demonstrated), so it'll stay!

**Comment by djerius** [comment]
Oh bother.  I've run into a problem with this approach.  Here's an example:

```
sub apply_roles {
    my @roles = @_;
    sub {
        my ( $bundles, $args ) = @_;
        Moo::Role->apply_role_to_package( $args->{package}, @roles );
        ();
      }
}

our %IMPORT_BUNDLES = (
    Class => [
        'Moo',
        apply_roles( 'App::acisraw::UndefTolerant' ),
    ],
);
```

The problem is that the roles are applied immediately, not in the order specified in the bundle list, so `Moo` doesn't get imported before the role is applied, and things break.  So, instead of acting like

```
use Moo;
with 'App::acisraw::UndefTolerant';
```

it acts like

```
with 'App::acisraw::UndefTolerant';
use Moo;
```
In this particular example, the role wraps `BUILDARGS`, and with the inverted application of a role, results in a 

```
The method 'BUILDARGS' is not found in the inheritance hierarchy for class [...]
```

error message.

**Comment by preaction** [comment]
That's weird. It looks like it should work, the ordering is preserved for exactly that reason.

You can try forcing Moo to be first, in front of everything, including `@IMPORT_MODULES`, which it usually should be anyway (because it also imports `strict` and `warnings` and would override any attempt by you to customize those), by prepending a `<`, like:

```
our %IMPORT_BUNDLES = (
    Class => [
        '<Moo',
        apply_roles( 'App::acisraw::UndefTolerant' ),
    ],
);
```

If that works, let me know. I'm still going to look into why what you posted did not work, because I think it should have.

The standard response to that error message, disregarding Import::Base, is to add a blank "sub BUILDARGS { }" to the role. It won't be imported if BUILDARGS exists in the class already, but if it does not, it will save you from seeing that error.

**Comment by djerius** [comment]
I think what's happening is that the subrefs are executed while the module lists are being created:
``` 
     57     while ( @modules ) {
     58         if ( ref $modules[0] eq 'CODE' ) {
     59             unshift @modules, shift( @modules )->( @cb_args );
     60             next;
     61         }
     62 
     63         my $module = shift @modules;
     64         if ( ref $modules[0] eq 'CODE' ) {
     65             unshift @modules, shift( @modules )->( @cb_args );
     66         }
  [...]
```
The modules aren't actually imported until later on, after the damage is done.

For example:

LoudMoo.pm
```
use 5.10.0;

package LoudMoo {
    sub import { say __PACKAGE__, ": MOO" }
}
1;
```

test.pl
```
use 5.10.0;

package Base {
    use parent 'Import::Base';
    our @IMPORT_MODULES = (
        'LoudMoo',
        sub { say "Quiet Moo"; () }
    );
}

Base->import;

```

The output:
```
% perl test.pl
Quiet Moo
LoudMoo: MOO
```

importing `<LoudMoo` instead of `LoudMoo` doesn't change the result.

**Comment by preaction** [comment]
Ah. Yes, that is what happens. Somehow I didn't get the e-mail about your reply. I'll try to come up with a way to address this if I can find one that doesn't suck. Right now "return a subref from the subref" doesn't sound too appealing...

I think what I do works because it doesn't matter the exact order/timing that @ISA gets altered.

**Comment by preaction** [comment]
Okay. Looks like I call the subrefs during construction of the list of modules, because the list of modules can include modules that need to be reordered. Since I want this exact use-case to work, I have to make a compromise: Subrefs cannot return classes that can be reordered with '<' and '>'.

I think being able to apply roles is much more useful than having a subref generate a dependency at runtime that is then shoved to the start of the list (when there are a few other ways to solve that which are more declarative and less magic).

**Comment by preaction** [comment]
v0.011 is uploaded to PAUSE and on its way to CPAN. Let me know if it works or not, and thanks for all the digging!

**Comment by djerius** [comment]
Hi,

Sorry for the late reply; lot's of snow on the US East Coast.  

I've run into another problem with applying roles which can't (as far as I know) be easily solved.  The problem is that when a role is applied during module import, it is done prior to compiling the importing package's code, so, at least in the case of `Moo` and `BUILDARGS`, `Moo::Role` creates a default one which collides with one declared in the importing module.  I've created an example and put a shell archive with all of the files here:

https://gist.github.com/djerius/bfb6af97640851fe10de

You'll see that Class2.pm won't compile (due to multiple `BUILDARGS`) while Class.pm will.  The difference is that in Class2, the role is applied prior to compilation of the class while in Class, the role is applied after compilation.  I keep stumbling over this. I wish I'd remember this subtlety! 

So, you can't get complete role application in an import.

**Comment by djerius** [comment]
A bit more sleuthing provided this: [Check::UnitCheck](https://metacpan.org/pod/Check::UnitCheck), which allows inserting `UNITCHECK` blocks into an importing package.  `UNITCHECK` blocks are run after the unit which defined them has been compiled.  Normally you can't inject them into another unit of code (which is what made me think this unsovable) but `Check::UnitCheck` allows that.  It's an XS module, unfortunately.

(More info on  `UNITCHECK` blocks is [here](http://perldoc.perl.org/perlmod.html#BEGIN%2c-UNITCHECK%2c-CHECK%2c-INIT-and-END).  `UNITCHECK` blocks arrived with Perl 5.10.0.)

So far, the following seems to work, but if history is any predictor, there's bound to be another corner case waiting...

````
package Base;

use parent 'Import::Base';

use Check::UnitCheck;
use Moo::Role;

sub apply_roles {
    my @roles = @_;
    sub {
        my ( $bundles, $args ) = @_;
        my $package = $args->{package};
        Check::UnitCheck::unitcheckify(
            sub {
                Moo::Role->apply_roles_to_package( $package, @roles );
                return;
            } );
      }
}

our @IMPORT_MODULES = (
    'Moo',
    apply_roles( 'Role' ),

);
1;
````

**Comment by preaction** [comment]
At that point, I might just say this is a bit out of scope of Import::Base.

But, perhaps I could add an API that would allow adding the bundles at runtime instead, which would solve a few more cases (like what if my "requires" is provided by an attribute, attributes are only added at runtime, so the requires won't be resolved when the role is applied)

Then, since we're in runtime, the full unit has been compiled and we're safe to do what we need to do.

Something perhaps like "My::Base->apply_bundle( 'BUNDLE_NAME' )" to only apply the named bundle? Ooh, yeah! Then I could "use My::Base" to get the base imports, and "My::Base->apply_bundle( 'Plugin' )" to make it a "Plugin". Not sure I like the name "apply_bundle", but I can bikeshed that until a good name floats to the top.

**Comment by djerius** [comment]
I agree it's a bit out of scope.  I feel a bit guilty about not exploring the whole issue prior to submitting this ticket, and since you've updated the docs to include an example of applying a role, I wanted to ensure that it would actually work as I advertised.

I think that API would work.  Perhaps allow `apply_bundle` to be exported for some sugar.

It's still annoying this has to be split into two steps, but that's the nature of the beast.

Thanks for your effort on this.

**Comment by preaction** [comment]
An export would be a bit nicer, but I'm already using the import list as bundles. So it would have to be like:

```
use My::Base { export => 'apply_bundle' };
apply_bundle 'Plugin';
```

Which now that I write it out, isn't as terrible as I feared.

Though, to think out loud, perhaps it should be `apply_base` instead:

```
use My::Base;
My::Base->apply( 'Plugin' );
```
```
use My::Base { export => 'apply_base' };
apply_base( 'Plugin' );
```

Or perhaps `with_bundle`?

**Comment by djerius** [comment]
How about

````
use My::Base -export => 'with_bundle', @bundles;
````

and allow for

````
use My::Base -export => [ 'with_bundle', 'other_routine' ], @bundles;
````

in case you add more exportable routines?  A bit more parsing, but less visual clutter for the user.

I like `with_bundle`.

**Comment by preaction** [comment]
Right, of course, I reserved a single dash for Import::Base-related arguments to import(), and double-dash for custom user arguments to import().

Alright. I like `use My::Base -import => 'with_bundle'; with_bundle @bundles;`. I'll do that. And then, yeah, if it's an arrayref, it allows multiple stuff. And then `My::Base->import_bundle( 'Plugin' );` is another option, which `with_bundle` is exactly `*{caller . "::with_bundle"} = sub { __PACKAGE__->import_bundle( @_ ) };`.

I'll probably change my mind about the color of the bikeshed by the time I get home to actually do this ;)

Copying this into a new ticket, since it's a new feature:  #23

**Comment by preaction** [comment]
I've added `My::Base->import_bundle` now in c555f13f86aa8180ba8028910a6ba5e71b7e1718. I'm going to add another example test using some Moo roles before I release it out, but that should be tonight.

I didn't make Import::Base into an exporter though, mainly because I found out exactly how few use-cases `import_bundle` really has (more on that in #23). But the one important use case it has is enough for it to exist. It just isn't enough for it to go all out on making it easy to use (unless and until it turns out that it's used more often than I think).
